contrib/colab/PyTorch_Bert_Multi_Core_TPU_Colab.ipynb

```diff
@@ -0,0 +1,865 @@
+        "        \n",
+        "        input_ids_q, input_masks_q, input_segments_q = [], [], []\n",
+        "        input_ids_a, input_masks_a, input_segments_a = [], [], []\n",
+        "        \n",
+        "        for _, instance in df[columns].iterrows():\n",
+        "            \n",
+        "            t, q, a = instance.question_title, instance.question_body, instance.answer\n",
+        "\n",
+        "            ids_q, masks_q, segments_q, ids_a, masks_a, segments_a = _convert_to_transformer_inputs(t, q, a, tokenizer, max_sequence_length)\n",
+        "            \n",
+        "            input_ids_q.append(ids_q)\n",
+        "            input_masks_q.append(masks_q)\n",
+        "            input_segments_q.append(segments_q)\n",
+        "\n",
+        "            input_ids_a.append(ids_a)\n",
+        "            input_masks_a.append(masks_a)\n",
+        "            input_segments_a.append(segments_a)\n",
+        "            \n",
+        "        return [\n",
+        "            np.asarray(input_ids_q, dtype=np.int32), \n",
+        "            np.asarray(input_masks_q, dtype=np.int32),\n",
+        "            np.asarray(input_segments_q, dtype=np.int32),\n",
+        "            np.asarray(input_ids_a, dtype=np.int32), \n",
+        "            np.asarray(input_masks_a, dtype=np.int32), \n",
+        "            np.asarray(input_segments_a, dtype=np.int32),\n",
+        "        ]\n",
+        "\n",
+        "    def compute_output_arrays(df, columns):\n",
+        "        return np.asarray(df[columns])\n",
+        "\n",
+        "    def compute_spearmanr_ignore_nan(trues, preds):\n",
+        "        \n",
+        "        rhos = []\n",
+        "        for tcol, pcol in zip(np.transpose(trues), np.transpose(preds)):\n",
+        "            rhos.append(spearmanr(tcol, pcol).correlation)\n",
+        "        return np.nanmean(rhos)\n",
+        "\n",
+        "    def train_model(train_loader, length, model, optimizer, criterion, scheduler=None):\n",
+        "        \n",
+        "        max_grad_norm = 1.0\n",
+        "        tracker = xm.RateTracker()\n",
+        "        avg_loss = 0.\n",
+        "        model.train();\n",
+        "        len_loader = length\n",
+        "        tk0 = enumerate(train_loader)\n",
+        "\n",
+        "        for idx, batch in tk0:\n",
+        "            \n",
+        "            input_ids_q, input_masks_q, input_segments_q, input_ids_a, input_masks_a, input_segments_a, labels = batch\n",
+        "            input_ids_q, input_masks_q, input_segments_q, input_ids_a, input_masks_a, input_segments_a, labels = input_ids_q.to(device, dtype=torch.long), input_masks_q.to(device, dtype = torch.long), input_segments_q.to(device, dtype = torch.long), input_ids_a.to(device, dtype = torch.long), input_masks_a.to(device, dtype = torch.long), input_segments_a.to(device, dtype = torch.long), labels.to(device, dtype = torch.float)\n",
```

Comment:
Tensors coming off the ParallelLoader are already on TPU device.

Reply:
Fixed!

Reply:
I think we had to do this otherwise the code with throw an error since Transformers isn't expecting XLAType Tensors as such isn't?

> RuntimeError: Expected tensor for argument #1 'indices' to have scalar type Long; but got XLAIntType instead (while checking arguments for embedding)